charms/istio-pilot/src/charm.py

```diff
@@ -293,23 +297,38 @@ def handle_default_gateway(self, event):
         # Update the ingress resources as they rely on the default_gateway
         self.handle_ingress(event)
 
-    def handle_gateway_info_relation(self, event):
-        if not self.model.relations["gateway-info"]:
-            self.log.info("No gateway-info relation found")
-            return
+    def handle_gateway_info_relation(self, _) -> None:
+        """Sends the default-gateway info through the gateway-info relation.
+
+        Raises:
+            GenericCharmRuntimeError("default-gateway not present in cluster):
+                if the default-gateway is not present in the cluster
+        """
+        # Verify the default-gateway is present in the cluster
         is_gateway_created = self._resource_handler.validate_resource_exist(
             resource_type=self._resource_handler.get_custom_resource_class_from_filename(
                 "gateway.yaml.j2"
             ),
             resource_name=self.model.config["default-gateway"],
             resource_namespace=self.model.name,
         )
-        if is_gateway_created:
-            self.gateway.send_gateway_relation_data(
-                self.app, self.model.config["default-gateway"], self.model.name
-            )
-        else:
+
+        # Raise if the default-gateway is not created, it must be present always
+        if not is_gateway_created:
             self.log.info("Gateway is not created yet. Skip sending gateway relation data.")
+            raise GenericCharmRuntimeError("default-gateway not present in cluster.")
+
+        # Send the default gateway name and namespace
+        try:
+            self.gateway_provider.send_gateway_relation_data(
+                gateway_name=self.model.config["default-gateway"],
+                gateway_namespace=self.model.name,
+            )
+        except GatewayRelationMissingError:
+            # If there is no relation, we cannot send data. In reality, the execution
+            # should not get this far.
+            self.unit.status = BlockedStatus("Please add gateway-info relation.")
+            return
```

Comment:
I don't think we need this here.  This charm provides gateway-info information to anyone who is listening.  If nobody is listening, nothing is blocked it just means a user has chosen not to use this function.

Reply:
But if you are observing the relation event and handling it, you should expect the relation (see also my comment one line above).

Reply:
You should expect the relation, but not expect that applications are related to it.  For example, doing:

```
juju deploy istio-pilot
juju deploy istio-gateway
juju relate istio-pilot istio-gateway
```
should result in istio-pilot and istio-gateway being green right?  But with this check, we'll actually go to BlockedStatus until we do:
```
juju deploy something-that-needs-gateway-info
juju relate istio-pilot:gateway-info something-that-needs-gateway-info:gateway-info
```

which I don't think is what we want.  

The problem is really stemming from [the Provider class](https://github.com/canonical/istio-operators/blob/b834927084bf26759813f9e2b86b40b23c05471b/charms/istio-pilot/lib/charms/istio_pilot/v0/istio_gateway_info.py#L230) which should not raise if there are no related applications, but instead should silently proceed, but we missed it in the other PR.

Reply:
You're right that I'll be touching this too in the next day or so so maybe its better for me to adjust it then, but with `GatewayRelationMissingError`s being raised in this situation the relation is not backward compatible (because istio-pilot will go to blocked in situations it used to not go to blocked for) and I'd argue semi-broken since basic deployments will go blocked.

Reply:
I might be misunderstanding this too, so if I am tell me I'm wrong :)  I'm a little confused why this hasn't broken the integration tests.  Although if I look at their [logs](https://github.com/canonical/istio-operators/actions/runs/4747082970/jobs/8431528979?pr=240), I'm not sure if they're passing or False negativing-ing

Reply:
The only case where the exception is raised is if you are using `send_gateway_relation_data` without having the relation, which is less likely because this event handler is reacting to the `gateway_info_relation_*` event.

In the best interest of time, I have changed this PR with minimal changes, you can decide how to proceed later.